text/0000-file-open-with.md

```diff
@@ -0,0 +1,116 @@
+- Feature Name: file_open_with
+- Start Date: 2018-11-08
+- RFC PR: 
+- Rust Issue: #55762
+
+# Summary
+[summary]: #summary
+
+This RFC proposes making `File::open()` consistent by deprecating `OpenOptions::new().read(true).write(true).open("existing_file")` and adding `File::open_with("existing_file", OpenOptions::new().read().write())` instead.
+
+# Motivation
+[motivation]: #motivation
+
+The current way to open an existing file in read-only mode is this:
+
+    File::open("foo.txt")
+
+And to create a new file it is:
+
+    File::create("foo.txt")
+
+But if you want to open an existing file in read/write mode it is this:
+
+use std::fs::OpenOptions;
+
+    let mut file = OpenOptions::new()
+        .read(true)
+        .write(true)
+        .open("foo.txt");
+
+This is inconsistent and unexpected. I propose that we deprecate `OpenOptions::open()` and add `File::open_with(path: &str, options: &OpenOptions)`.
+
+    let mut file = File::open_with("foo.txt", OpenOptions::new().read().write());
+
+This matches the normal way of doing "call a method with some options" in Rust, for example `TcpStream::connect(addr)` and `TcpStream::connect_timeout(addr, timeout)`.
```

Comment:
The improvement is discoverability and consistency. There's already `File::open()` and `File::create()` which leads users to believe that to open files you use a method on `File`.

If `OpenOptions` had been named `FileOpenBuilder` or something then I agree it would have been a bit clearer but given that it isn't, *and* that the most obvious methods to open a file are on `File` it seems silly that some random subset of file open methods are on `OpenOptions`.

This may be one of those things that feels obvious once you have learnt it, but as a Rust beginner, trust me this was weird and confusing. (And see the Stackoverflow questions for further evidence.)

Reply:
Perhaps then the docs on `File::{open, create}` should be improved to mention that you need to use `OpenOptions` to configure permissions

Reply:
That would certainly be a good thing to do, however it would be better if the API were intuitive in the first place. Nobody likes reading documentation.

Reply:
For the specific use case of opening a file, I don't think the builder pattern makes much intuitive sense. After all, you're not building a file. You're building an `OpenOptions` object. Or perhaps you could say you're building a file descriptor. But it seems weird to say that the file itself has been "built" before we write any content into it. Maybe it's possible to design a useful `FileBuilder` API where `.build()` actually puts content in the file, but that'd be a layer above what `std` is trying to provide.

I think it's clear that `File::open(name, options)` is a much more intuitive API. To me the only question is whether adding it and deprecating the already stable alternative is a net win for the language and its ecosystem, which is where I'll have to defer to others who know the Rust ecosystem much better.